### ISSUE TITLE
Significant canvas performance degradation in v2.0.x

### ISSUE THREAD

**Issue by MaxGraey** [electron/electron#12042]
* Electron version: v2.0.0-beta.1
* Operating system: Mac OSX: 10.13.3

### Expected behavior

Performance similar to previous versions (especially v1.6.17)

### Actual behavior

Slowdown approximately in 5-8x times

### How to reproduce

- Install electron v2.0.0-beta.1
- Launch any canvas app like this for example:

```html
<!DOCTYPE html>
<html>

  <head>
    <script src="https://cdn.rawgit.com/konvajs/konva/1.7.6/konva.min.js"></script>
    <meta charset="utf-8">
    <title>Konva Optimizing Strokes Demo</title>
    <style>
      body {
        margin: 0;
        padding: 0;
        overflow: hidden;
        background-color: #F0F0F0;
      }
    </style>
  </head>

  <body>
    <div id="container"></div>
    <script>
      var width = window.innerWidth;
      var height = window.innerHeight;

      var stage = new Konva.Stage({
        container: 'container',
        width: width,
        height: height
      });
      var layer = new Konva.Layer();
      stage.add(layer);

      var n = 100, shape;
      for (var i = 0; i < n; i++) {
        shape = new Konva.Circle({
          x: stage.width() * Math.random(),
          y: stage.height() * Math.random(),
          radius: 10 + 10 * Math.random(),
          fill: Konva.Util.getRandomColor(),
          stroke: 'black',
          shadowEnabled: false,
          draggable: true,
        });
        shape.strokeHitEnabled(false);
        shape.shadowForStrokeEnabled(false);

        layer.add(shape);
      }

      layer.draw();
    </script>

  </body>

</html>
```
- drag some circle

**Comment by MaxGraey** [comment]
It seems problem with new color correction stuffs which improve color banding in gradients.
Solution for Chrome v61: https://stackoverflow.com/a/46167982
For Chrome >= 62: https://stackoverflow.com/a/47165506

**Comment by bpasero** [comment]
@Tyriar isn't this https://github.com/electron/electron/issues/11051 ?

**Comment by MarshallOfSound** [comment]
@Tyriar If it is can you check those workarounds 😄 Might be worth documenting or something 🤔

**Comment by Tyriar** [comment]
Well https://github.com/electron/electron/issues/11051 was specifically on Linux, OP is on mac. Also I thought I verified this issue on v2 but I only checked on mac 😄

**Comment by alexeykuzmin** [comment]
@MaxGraey

> **Expected behavior**
> Performance similar to previous versions (especially v1.6.17)

Is there any difference between 2.0.0-beta1 and the latest stable Electron 1.8.2?

**Comment by MaxGraey** [comment]
> Is there any difference between 2.0.0-beta1 and the latest stable Electron 1.8.2?
Yes, as well. But difference between 2.0.0-beta1 and 1.6.x/1.7.x is much wider. I just mentioned that canvas perf for 1.6.17 the best on OSX in my opinion.

**Comment by MaxGraey** [comment]
For 1.6.x/1.7.x:

```js
getGPUFeatureStatus { '2d_canvas': 'enabled',
  flash_3d: 'enabled',
  flash_stage3d: 'enabled',
  flash_stage3d_baseline: 'enabled',
  gpu_compositing: 'enabled',
  multiple_raster_threads: 'enabled_on',
  native_gpu_memory_buffers: 'enabled',
  rasterization: 'enabled',
  video_decode: 'enabled',
  video_encode: 'enabled',
  vpx_decode: 'enabled',
  webgl: 'enabled',
  webgl2: 'enabled' }
```

for >= 1.8.x
```js
getGPUFeatureStatus { '2d_canvas': 'unavailable_software',
  flash_3d: 'enabled',
  flash_stage3d: 'enabled',
  flash_stage3d_baseline: 'enabled',
  gpu_compositing: 'enabled',
  multiple_raster_threads: 'enabled_on',
  native_gpu_memory_buffers: 'enabled',
  rasterization: 'unavailable_software',
  video_decode: 'enabled',
  video_encode: 'enabled',
  webgl: 'enabled',
  webgl2: 'enabled' }
```
So we need 
turn on ```ignore-gpu-blacklist``` or enable ```enable-accelerated-2d-canvas```/```enable-gpu-rasterization``` and ```enable-native-gpu-memory-buffers```

But this performance lower than 1.6.x/1.7.x any more.

**Comment by alexeykuzmin** [comment]
JFYI `gpu_rasterization` has been disabled on Mac >= 10.13 as a workaround for "[Rendering glitches in High Sierra (macOS 10.13)](https://bugs.chromium.org/p/chromium/issues/detail?id=773705)".

**Comment by MaxGraey** [comment]
I haven't this issue, but I resetting NVRAM more recently.

**Comment by Tyriar** [comment]
@npezza93 oh? I guess I'll have to try again, the main thing I checked for was 60ish fps in the terminal. Can you give some approximations of the numbers you were seeing?

**Comment by npezza93** [comment]
On 1.8.x
<img width="1552" alt="screen shot 2018-03-08 at 7 34 33 am" src="https://user-images.githubusercontent.com/5059927/37151385-4ac40708-22a3-11e8-9d4d-8e14912c08eb.png">

On 2.0 beta
<img width="1307" alt="screen shot 2018-03-08 at 7 35 55 am" src="https://user-images.githubusercontent.com/5059927/37151400-5c1d22dc-22a3-11e8-8340-db174723721f.png">


Both of the big blocks I ran the same command in xterm (I cat'd a yarn.lock file). In 1.8 it took about 3 seconds in 2.0 it took a little over double that.

**Comment by npezza93** [comment]
Hey @Tyriar, some more numbers and some steps to repro:

Using the example electron app inside node-pty(https://github.com/Tyriar/node-pty/tree/master/examples/electron), I ran `cat node_modules/xterm/package-lock.json` from inside the example app. I used electron versions 1.7.11 and 2.0.0-beta.2 and saw the following results. 


v1.7.11
<img width="1446" alt="screen shot 2018-03-08 at 5 45 19 pm" src="https://user-images.githubusercontent.com/5059927/37180961-03418b3e-22f9-11e8-9a6a-4c1db92f8cc5.png">

v2
<img width="1841" alt="screen shot 2018-03-08 at 5 38 35 pm" src="https://user-images.githubusercontent.com/5059927/37180942-e9614da8-22f8-11e8-9ef4-1bce6b7fc80a.png">

**Comment by ckerr** [comment]
This can't be fixed until c63. There are switches that apps can use to speed things up; however some may have side-effects and so they can't be done unilaterally in every case.

@MarshallOfSound has been volunteered :beer: to write up documentation for these options for the 2.0.0 release

**Comment by MarshallOfSound** [comment]
@npezza93 Can you try your repro running electron with `--ignore-gpu-blacklist`

**Comment by npezza93** [comment]
Without flag:
<img width="1552" alt="screen shot 2018-03-12 at 5 07 56 pm" src="https://user-images.githubusercontent.com/5059927/37309780-504ef764-2618-11e8-97f6-b2d738238b57.png">
With flag:
<img width="1552" alt="screen shot 2018-03-12 at 5 09 22 pm" src="https://user-images.githubusercontent.com/5059927/37309786-557ec0ac-2618-11e8-8cb5-e2198f19ee6f.png">

Worked like a champ! Thanks, @MarshallOfSound!

**Comment by MarshallOfSound** [comment]
@npezza93 Good to hear we identified the issue, but our main point still stands.  This won't be properly fixed for a while.  By using this flag you are telling Chromium to ignore the GPU blacklist it uses to determine feature stability on certain OS's / GPU's.  The thing that will affect you the most is graphical artifacts appears / graphical glitches on macOS high sierra and nvidia GPU's.  If it's 100% required use the flag, but if you can avoid it and take the performance hit until Apple fix the nvidia drivers on high sierra I'd recommend doing that 👍

**Comment by ckerr** [comment]
Marking this as `blocked-on-upstream` for now, as there's not a lot that can be done in Electron until this is fixed upstream.

**Comment by bpasero** [comment]
@MarshallOfSound @ckerr is this the same consequence as applying https://github.com/electron/electron/pull/10898 for 1.7.x? VS Code was already running with that patch for a while now and I would like to understand if the performance regression originates from there or if this is something new.

**Comment by Tyriar** [comment]
@bpasero I do not see a particular change in performance between our 1.7 and 2.0 VS Code builds, that may be because we've tweaked flags already though?

**Comment by MarshallOfSound** [comment]
@bpasero Afaik the performance regression stems from that patch (or it's counterpart in chromium)

@Tyriar It is also hardware specific, depends on the GPU in your mac 👍

**Comment by Tyriar** [comment]
@MarshallOfSound this sounds a lot like https://github.com/electron/electron/pull/10923, did that patch get reverting while upleveling Chromium?

**Comment by bpasero** [comment]
It looks like this patch still exists in Electron 2.0.x (proof: https://github.com/electron/libchromiumcontent/blob/electron-2-0-x/patches/043-fix_rendering_glitches_on_high_sierra.patch) because the related Chrome issue was fixed after Chrome 61 (possibly Chrome 62).

VSCode always had this patch, we backported it early on to 1.7.x.

**Comment by Tyriar** [comment]
I actually am seeing this on Linux. Here's the output of getGPUFeatureStatus:

```
{ '2d_canvas': 'unavailable_software',
  checker_imaging: 'disabled_off',
  flash_3d: 'unavailable_software',
  flash_stage3d: 'unavailable_software',
  flash_stage3d_baseline: 'unavailable_software',
  gpu_compositing: 'unavailable_software',
  multiple_raster_threads: 'enabled_on',
  native_gpu_memory_buffers: 'disabled_software',
  rasterization: 'unavailable_software',
  video_decode: 'unavailable_software',
  video_encode: 'unavailable_software',
  webgl: 'enabled_readback',
  webgl2: 'unavailable_off' }
```

Running with command line flags fixes the issue:

```
./scripts/code-cli.sh --enable-accelerated-2d-canvas --enable-gpu-rasterization --ignore-gpu-blacklist --enable-native-gpu-memory-buffers
```

**Comment by ckerr** [comment]
Now that https://github.com/electron/libchromiumcontent/issues/489 has landed, this may be ameliorated in 2.0.0-beta.6 which will be out in a day or two

**Comment by Tyriar** [comment]
FYI I think electron should revert https://github.com/electron/libchromiumcontent/pull/489 as removing this version of macOS from the blacklist reintroduces the square artifacts https://github.com/Microsoft/vscode/issues/48043

VS Code is solving the problem by adding a DOM fallback to the canvas implementations. There will always be environments that have been blacklisted by Chromium for various reasons (like my work Linux desktop), so we needed a general solution.

**Comment by bpasero** [comment]
I think we cannot ship with a 2.0 that brings back the ugly square artefacts. So I am +1 for reverting the GPU blacklist change if that causes it.

**Comment by bpasero** [comment]
@Tyriar @deepak1556 it looks like Chrome is stating that the rendering issues have been fixed in macOS 10.13.4 and so their latest patch is actually going even further by blacklisting only if macOS < 10.13.4 (via [commit](https://chromium.googlesource.com/chromium/src.git/+/e9b6fe054991cc76c4a67c4052949be63d1b1c73)).

**Comment by bpasero** [comment]
Unfortunately at least one [user confirmed](https://github.com/Microsoft/vscode/issues/48043#issuecomment-396000273) that the background artefacts are **still happening** even with latest `macOS 10.13.5` for him.

**Comment by MaxGraey** [comment]
I should mentioned this problem resolved in v3.0.0-beta.1 with Chromium 66 for MacOS.
